### DISCUSSION TITLE
v6 dynamic segment with enumerated values

### DISCUSSION THREAD

**Discussion by JakobJingleheimer** [remix-run/react-router#10951] [Q&A]
I want to use route params for (among other things) determining API requests and also specify a component for each valid value. The following works in terms of route params and mounting the layout component `SectionContainer` (which contains an `<Outlet />`), but the appropriate route component is not loaded.

```jsx
<Route
  path=":section"
  Component={SectionContainer}
  loader={({ params: { section } }) => {
    const endpoint = endpoints[section];
    const data = be.get(endpoint);
    return defer({ data });
  }}
>
  <Route
    index
    Component={Foo}
  />
  <Route
    path="bar"
    Component={Bar}
  />
  <Route
    path="qux/:subSection"
    Component={Qux}
  >
    <Route
      path="zed"
      Component={Zed}
    />
    {/* … */}
  </Route>
</Route>
```

I've tried a couple variations of this (ex adding a "wrapper" `<Route path=":section">` and nesting `<Route Component={SectionContainer}>` within), but everything I try has the same result: `Matched leaf route at location "…/foo" does not have an element or Component. […]`

I feel like this should be possible, I'm just not finding the right mix.

P.S. Sorry if this has already been answered—I looked through discussions, issues, and the docs, but I didn't find anything.

EDIT: repro https://codesandbox.io/s/react-router-v6-dynamic-param-enums-xs9ncr?file=/src/index.js

**Comment by kiliman** [comment]
Can you provide a minimal reproduction?

**Reply by JakobJingleheimer** [reply]
Is there a code sandbox or something? (The repro is pretty simple)

I have a workaround, basically re-implementing useParams. But I would rather do something blessed.

**Reply by JakobJingleheimer** [reply]
Repro: https://codesandbox.io/s/react-router-v6-dynamic-param-enums-xs9ncr?file=/src/index.js

in the preview, set the pathname to `foo` and then change it to `bar` to see the useParams value correctly updating but the wrong component used (it's always `<Foo>`).

**Reply by kiliman** [reply]
I think I understand the problem now. You're expecting the `:section` param, to match `foo|bar|qux`. That's not how it works. The `<SectionContainer>` route will *grab* the first segment as the param `:section`. The child routes will then match against the rest of the URL. React Router doesn't support dynamic route matching.

So if you do `/test` it will display `<Bar>`. `/test/foo` will display `<Foo>`, etc. As you can see, you still need the initial `:section` param, before it matches.

`<Qux>` was missing an `<Outlet>` for `<Zed>`.

So the URL `/test/qux/test2/zed` displays like this:

<img width="412" alt="image" src="https://github.com/remix-run/react-router/assets/47168/5c1f7e7c-c5fa-4672-9e2d-d94065120564">

If you really need to be able call `foo`, `bar`, etc endpoints, you can still do that. React Router will match the correct components, and you'll be able to access the parent loader's data via `useMatches`.

```ts
const routes = [
  {
    Component: App,
    children: [
      {
        path: '/', // remove :section param
        Component: Container,
        loader({ request }) {
          // get first segment of URL as section name (default to foo)
          // skip leading '/'
          const section = request.pathname.substring(1).split('/')[0] ?? 'foo'

          const endpoint = endpoints[section];
          const data = be.get(endpoint);
          return { data };
        },
```

**Reply by kiliman** [reply]
That said, I recommend you move your loader code to each route component. That way, you know exactly what data is being used by each route.

**Reply by JakobJingleheimer** [reply]
Mm, I need to access params elsewhere in the app, so splitting pathname all over the place is not really a viable solution.

I had thought if I made `:section` (and `:subSection`) optional, react-router might do both.

Would you consider a feature request for this functionality? I've been doing this since AngularJS v1. It's very handy.

**Reply by JakobJingleheimer** [reply]
> That said, I recommend you move your loader code to each route component. That way, you know exactly what data is being used by each route.

Do you mean into the route component's file? IRL, I do and use `lazy()`; for the sample here I put it inline for simplicity.

**Reply by kiliman** [reply]
I mean, `<Foo>` should have a loader calling the `foo` endpoint, etc. Instead of having a parent route that does that.

**Reply by JakobJingleheimer** [reply]
Ah, that is a good point. If the child loader can access its parent's data, then that works great actually. The problem was that the parent (container) fetches a generic subset of the data (that all children need, and that part of the UI renders whilst the child is sorting itself out) and then children fetch a different subset specific to the child (but they're from the same entity—say Product). So if I can merge them together before returning to the child, win-win.
